File: app/calculo/secoes/periodo.py
# -*- coding: utf-8 -*-
"""Análise por período — porte de renderPeriodo(), PRESETS e buildFilter() (app.js). Parâmetros: de, ate."""
from ..datas import ym_shift
from ..numeros import ordem_chaves_js
from ..vendas import PSEUDO_VEND, agregar, tabela_clientes

CLSMAP = {'CLIENTE': 'Cliente final', 'SHOWROOM': 'Showroom', 'SHOW ROOM': 'Showroom', 'Não informado': 'Não informado',
          'BONIFICACAO': 'Bonificação', 'CANCELADO': 'Cancelado'}


def _int(v, padrao):
    try:
        return int(v)
    except (TypeError, ValueError):
        return padrao


def filtro(C, params):
    valido = lambda v: v in (200001, 300000) or (200001 <= v <= 300000 and 1 <= v % 100 <= 12)
    ini = C.ano_v * 100 + 1
    a, b = _int(params.get('de'), ini), _int(params.get('ate'), C.maxym_vendas)
    a = a if valido(a) else ini
    b = b if valido(b) else C.maxym_vendas
    return (b, a) if a > b else (a, b)


def _carteira_hist(C, a_monthly):
    CH = (C.blob('DATA').get('carteira') or {}).get('hist')
    yms = sorted(a_monthly)
    if not (CH and CH['yms'] and len(yms) > 1):
        return None
    ym_a, ym_b = yms[0], yms[-1]
    hi = [i for i, y in enumerate(CH['yms']) if ym_a <= y <= ym_b]
    if len(hi) <= 1:
        return None
    DRE = C.blob('DRE') or {}
    hyms = [CH['yms'][i] for i in hi]
    hval = [CH['valor'][i] for i in hi]

    def mes_dre(ym):
        y, m = ym // 100, ym % 100 - 1
        d = DRE.get(str(y))
        return (d['months'][m] if m < len(d['months']) else 0) or 0 if d else 0

    def fat6m(ym):
        s = n = 0
        for k in range(6):
            v = mes_dre(ym_shift(ym, -k))
            if v:
                s += v
                n += 1
        return s / n if n else 0

    fat = [fat6m(y) for y in hyms]
    return {'yms': hyms, 'valor': hval, 'meses': [hval[i] / f if f > 0 else 0 for i, f in enumerate(fat)],
            'venda': [a_monthly.get(y, 0) for y in hyms], 'fat6m': fat}


def calcular(C, params=None):
    params = params or {}
    DATA = C.blob('DATA')
    a, b = filtro(C, params)
    g = agregar(DATA, a, b)
    out = {'periodo.abertura': {'a': a, 'b': b, 'ano_v': C.ano_v, 'maxym': C.maxym_vendas, 'minym': C.minym_vendas,
                                'anos_fechados': C.anos_fechados(C.ano_v),
                                'total': g['total'], 'peds': g['peds'], 'ticket': g['ticket'], 'qnt': g['qnt']}}
    ch = _carteira_hist(C, g['monthly'])
    if ch:
        out['pe-cart'] = ch
    vr = [x for x in g['vend'] if x['peds'] >= 1 and x['name'] not in PSEUDO_VEND]
    out['pe-vend'] = {'top': [[x['name'], x['v']] for x in vr[:10]],
                      'linhas': [[x['name'], x['v'], x['contrib'], x['peds'], x['ticket'], x['titem'], x['com'], x['custop']]
                                 for x in g['vend'] if x['v'] > 0 and x['name'] not in PSEUDO_VEND]}
    fams = [f for f in g['fams'] if f['name'] != 'Frete/Serviço']
    fam_tot = 0
    for f in fams:
        fam_tot += f['v']
    out['pe-cat'] = {'top': [[x['name'], x['v']] for x in fams[:12]], 'total': fam_tot, 'n': len(fams)}
    top_v = [x['name'] for x in vr[:6]]
    top_f = [x['name'] for x in fams[:8]]
    md = {vn: {fn: 0 for fn in top_f} for vn in top_v}
    for vi in ordem_chaves_js(g['VF']):
        vn = DATA['vd'][vi]
        if vn not in md:
            continue
        for fi in ordem_chaves_js(g['VF'][vi]):
            fn = DATA['fam'][fi]
            if fn in md[vn]:
                md[vn][fn] = g['VF'][vi][fi]
    out['pe-catvend'] = {'linhas': top_v, 'colunas': top_f, 'mat': [[md[vn][fn] for fn in top_f] for vn in top_v]}
    cagg = {}
    for x in g['clss']:
        k = CLSMAP.get(x['name'], x['name'])
        cagg[k] = cagg.get(k, 0) + x['v']
    cd = sorted(([k, cagg[k]] for k in ordem_chaves_js(cagg) if cagg[k] > 0), key=lambda e: -e[1])[:3]
    out['pe-canal'] = {'top': [[x['name'], x['v']] for x in g['clis'][:12]], 'canal': cd}
    tc = tabela_clientes(g['clis'], g['total'], maximo=25)
    tc['linhas'] = [{'name': x['name'], 'v': x['v']} for x in tc['linhas']]
    out['pe-cli'] = dict(tc, n_clis=len(g['clis']))
    out['pe-desig'] = {'top': [[x['name'], x['roy']] for x in g['desg'][:12]],
                       'linhas': [[x['name'], x['roy'], x['v'], x['roy'] / x['v'] if x['v'] else 0, x['peds']] for x in g['desg'][:15]]}
    out['pe-arq'] = {'top': [[x['name'], x['rt']] for x in g['arqs'][:12]],
                     'linhas': [[x['name'], x['rt'], x['v'], x['rt'] / x['v'] if x['v'] else 0, x['peds']] for x in g['arqs'][:15]]}
    c = g['comp']
    comp = [['RT arquitetos', c['rt']], ['Royalties designers', c['roy']], ['Comissão vendedor(a) (PF)', c['pf']],
            ['Prêmio', c['prem']], ['Comissão vendedor(a) (PJ)', c['pj']], ['Curadoria / RP', c['cur']], ['DSR', c['dsr']],
            ['Comissão gerente', c['ger']]]
    comp = sorted((x for x in comp if x[1] > 0), key=lambda x: -x[1])
    rep = 0
    for x in comp:
        rep += x[1]
    out['pe-repasses'] = {'comp': [[x[0], x[1], x[1] / g['total'] if g['total'] else None] for x in comp], 'total': rep,
                          'pct': rep / g['total'] if g['total'] else None}
    return out

File: app/static/relatorio/secoes/periodo.js
/* Análise por período — desenho de buildFilter/renderPeriodo do Kit.
   Presets e "Aplicar" pedem os números ao servidor (ctx.buscar). */
'use strict';
window.DESENHO=window.DESENHO||{};
window.DESENHO_POS=window.DESENHO_POS||{};

function _pePeriodo(A,a,b){ if(a<=200001&&b>=300000)return 'histórico completo ('+ymLabAno(A.minym)+'–'+ymLab(A.maxym)+')';
  if(a===b) return ymLab(a).replace('/','/20'); const Y1=Math.floor(a/100),Y2=Math.floor(b/100);
  if(a%100===1&&b%100===12&&Y1===Y2) return 'ano '+Y1; return ymLab(a)+' – '+ymLab(b);}

function _peCorpo(P){
  let s='';
  const A=P['periodo.abertura'];
  if(A) s+='<div class="kpis k4">'+kpi('Venda contratada',mi(A.total),_pePeriodo(A,A.a,A.b))
    +kpi('Pedidos',nf(A.peds),'')+kpi('Ticket médio',mi(A.ticket),'por pedido')
    +kpi('Quantidade',nf(A.qnt),'itens')+'</div>';
  const CH=P['pe-cart'];
  if(CH){
    s+=H3('Carteira em aberto — posição mês a mês','saldo reconstruído da data do pedido e da NF','pe-cart');
    s+='<div id="cart-hist-chart">'+fig(evolCombo(CH.yms.map(ymLab),[{name:'Carteira em aberto',values:CH.valor,color:SER[0]}],CH.meses,
      {valfmt:v=>mi(v,1),acumfmt:v=>nf(v,1)+'x',dashName:'Meses de carteira',dashArea:false,dashColor:BAD,dashLabels:true,
       leftLine:[{name:'Venda contratada',values:CH.venda,color:'#c9a227'},
                 {name:'Média faturamento 6M',values:CH.fat6m,color:SER[2],cls:'evolcombo-fat6m'}],
       linesHidden:true,w:980,h:320}),ins('carteiraMesesCobertura'))+'</div>';
    s+=cap('Barras = saldo em aberto no fim de cada mês (pedidos sem NF até aquela data — mesmo critério da foto atual de carteira). Linha amarela = venda contratada do mês (mesmo eixo, R$). Linha tracejada vermelha (eixo direito) = carteira ÷ faturamento médio dos últimos 6 meses (Receita Bruta da DRE) — quantas vezes o faturamento a carteira representa naquele mês.');
  }
  const V=P['pe-vend'];
  if(V){
    s+=H3('Desempenho por vendedor(a)','','pe-vend');
    s+=fig(hbar(V.top,{valfmt:v=>mi(v,1),maxbars:10}),ins('auto'));
    s+=table(['Vendedor(a)','Venda','% contrib.','Pedidos','Ticket/pedido','Ticket/item','Custo comissão','% custo'],
      V.linhas.map(x=>[esc(x[0]),money(x[1]),pct(x[2]),nf(x[3]),money(x[4]),money(x[5]),money(x[6]),pct(x[7])]),
      ['left','right','right','right','right','right','right','right'],null,ins('auto'));
  }
  const F=P['pe-cat'];
  if(F){
    s+=H3('Categorias (famílias de produto)','','pe-cat');
    s+=fig(hbar(F.top,{valfmt:v=>mi(v,1),color:SER[2],maxbars:12,pct:true,total:F.total}),ins('auto'));
    s+=cap('Percentuais sobre '+mi(F.total)+' — o total das categorias no período selecionado, já sem Frete/Serviço.'
      +(F.n>12?' O gráfico mostra as 12 maiores das '+F.n+' categorias, por isso o acumulado não fecha 100%.':''));
  }
  const M=P['pe-catvend'];
  if(M){ const md={}; M.linhas.forEach((vn,i)=>{md[vn]={}; M.colunas.forEach((fn,j)=>md[vn][fn]=M.mat[i][j]);});
    s+=H3('Principais categorias por vendedor(a)','R$ mil','pe-catvend');
    s+=fig(heatmap(M.linhas,M.colunas,md,{valfmt:v=>v>1000?nf(v/1000,0):'',padLeft:150,w:900}),ins('mixCruzado'));
  }
  const K=P['pe-canal'];
  if(K){
    s+=H3('Clientes e canal','','pe-canal');
    s+='<div class="two"><div>'+fig(hbar(K.top,{valfmt:v=>mi(v,1),padLeft:210,w:520,maxbars:12}),ins('auto'))+'</div>';
    s+='<div>'+fig(donut(K.canal,{valfmt:v=>mi(v,1)}),ins('auto'))+'</div></div>';
  }
  const T=P['pe-cli'];
  if(T){ const total=T.total; let acum=0;
    const trs=T.linhas.map((x,i)=>{ acum+=x.v;
      const ehCorte=T.corte&&i===T.corte.n-1;
      const cls=['cli-row',ehCorte?'cut':''].filter(Boolean).join(' ');
      return `<tr class="${cls}"><td class="left">${esc(x.name)}</td>`
        +`<td class="right">${money(x.v)}</td>`
        +`<td class="right">${pct(total?x.v/total:0)}</td>`
        +`<td class="right acum">${pct(total?acum/total:0)}</td></tr>`;}).join('');
    s+=H3('Maiores clientes do período',
      T.truncado?(nf(T.n)+' maiores de '+nf(T.n_clis)):
      T.corte?(nf(T.corte.n)+(T.corte.n===1?' cliente':' clientes')+' = 80% da venda'):'','pe-cli');
    s+='<div class="tw"><table class="dt"><thead><tr><th class="left">Cliente</th>'
      +`<th class="right">Venda</th><th class="right">${esc('% do total')}</th><th class="right">% acumulado</th>`
      +`</tr></thead><tbody>${trs}</tbody></table></div>`;
    s+='<div class="tw-ins"></div>';
    s+=cap(T.truncado
      ? 'Os '+nf(T.n)+' maiores clientes do período, que somam '+pct(T.share,1)+' da venda. Chegar a 80% exigiria '
        +nf(T.corte.n)+' clientes dos '+nf(T.n_clis)+' do período — quanto mais longo o período escolhido, mais espalhada fica a venda entre os clientes.'
      : (T.corte&&T.n>T.corte.n?'A linha marcada é onde o acumulado cruza os 80%. ':'')
        +(T.restantes>0?'Outros '+nf(T.restantes)+' clientes compraram no período e somam '+pct(1-T.share,1)+'.':'Todos os clientes do período estão na tabela.'));
  }
  const D=P['pe-desig'];
  if(D){
    s+=H3('Designers — royalties pagos','','pe-desig');
    s+=fig(hbar(D.top,{valfmt:v=>money(v),color:SER[4],padLeft:240,w:760,maxbars:12}),ins('auto'));
    s+=table(['Designer','Royalties pagos','Venda associada','% roy.','Pedidos'],
      D.linhas.map(x=>[esc(x[0]),money(x[1]),money(x[2]),pct(x[3]),nf(x[4])]),['left','right','right','right','right'],null,ins('auto'));
  }
  const R=P['pe-arq'];
  if(R){
    s+=H3('Arquitetos — RT paga','','pe-arq');
    s+=fig(hbar(R.top,{valfmt:v=>money(v),color:SER[6],padLeft:240,w:760,maxbars:12}),ins('auto'));
    s+=table(['Arquiteto / escritório','RT paga','Venda associada','% RT','Pedidos'],
      R.linhas.map(x=>[esc(x[0]),money(x[1]),money(x[2]),pct(x[3]),nf(x[4])]),['left','right','right','right','right'],null,ins('auto'));
  }
  const Q=P['pe-repasses'];
  if(Q){
    s+=H3('Composição de repasses e comissões','','pe-repasses');
    s+=fig(stackbar(Q.comp.map(x=>[x[0],x[1]]),{valfmt:v=>mi(v,1)}),ins('auto'));
    s+=table(['Componente','Valor pago','% das vendas'],Q.comp.map(x=>[esc(x[0]),money(x[1]),pct(x[2]==null?NaN:x[2])]),['left','right','right'],
      ['Total repasses',money(Q.total),pct(Q.pct==null?NaN:Q.pct)],ins('auto'));
  }
  return s;
}

DESENHO.periodo=function(P){
  const A=P['periodo.abertura'];
  let s='';
  if(A){
    const M=A.maxym, AV=A.ano_v;
    const presets=[['Histórico completo',200001,300000],
      ...A.anos_fechados.map(y=>[String(y),y*100+1,y*100+12]),
      ['YTD '+(AV-1)+' ('+ytdLab(M)+')',(AV-1)*100+1,M-100],
      [AV+' YTD ('+ytdLab(M)+')',AV*100+1,M],
      [MES_LONGO[M%100-1]+' '+Math.floor(M/100),M,M]];
    const opts=ymList(A.minym,A.maxym).map(y=>`<option value="${y}">${ymLab(y)}</option>`).join('');
    s+=`<div class="sec-head"><div class="kick">Exploração</div><h2>Análise por período</h2>
      <p class="lead">Selecione um período para recalcular todas as análises abaixo (vendedoras, categorias, clientes, designers, arquitetos e repasses).</p></div>
      <div class="filterbar"><div class="fb-presets">`
      +presets.map(p=>`<button class="preset" data-a="${p[1]}" data-b="${p[2]}">${esc(p[0])}</button>`).join('')
      +`</div><div class="fb-custom">De <select id="f-de">${opts}</select> até <select id="f-ate">${opts}</select>
      <button id="f-apply">Aplicar</button></div><div id="periodo-label" class="fb-label">Período: ${_pePeriodo(A,A.a,A.b)}</div></div>`;
  }
  s+='<div id="periodo-body">'+_peCorpo(P)+'</div>';
  return s;
};

DESENHO_POS.periodo=function(sec,ctx){
  const A=ctx.payload['periodo.abertura']; if(!A) return;
  const f={de:A.a,ate:A.b};
  const de=document.getElementById('f-de'), ate=document.getElementById('f-ate');
  const marca=()=>sec.querySelectorAll('.preset').forEach(el=>el.classList.toggle('on',+el.dataset.a===f.de&&+el.dataset.b===f.ate));
  async function corpo(){
    const P=await ctx.buscar(f);
    document.getElementById('periodo-body').innerHTML=_peCorpo(P);
    document.getElementById('periodo-label').textContent='Período: '+_pePeriodo(A,f.de,f.ate);
    marca(); ctx.remarcar();
  }
  sec.querySelectorAll('.preset').forEach(el=>el.addEventListener('click',()=>{f.de=+el.dataset.a; f.ate=+el.dataset.b;
    de.value=Math.max(A.minym,f.de===200001?A.minym:f.de); ate.value=Math.min(A.maxym,f.ate===300000?A.maxym:f.ate); corpo();}));
  document.getElementById('f-apply').addEventListener('click',()=>{let a=+de.value,b=+ate.value; if(a>b){const t=a;a=b;b=t;} f.de=a; f.ate=b; corpo();});
  de.value=f.de; ate.value=f.ate;
  marca();
};
